nvbench/iteration_space_base.cuh

```diff
@@ -0,0 +1,113 @@
+/*
+ *  Copyright 2022 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <nvbench/detail/axes_iterator.cuh>
+
+namespace nvbench
+{
+
+/*!
+ * Base class for all axi and axes iteration spaces.
+ *
+ * If we consider an axi to be a container of values, iteration_spaces
+ * would be how we can create iterators over that container.
+ *
+ * With that in mind we get the following mapping:
+ * * linear_axis_space is equivalant to a forward iterator.
+ *
+ * * zip_axis_space is equivalant to a zip iterator.
+ *
+ * * user_axis_space is equivalant to a transform iterator.
+ *
+ * The `nvbench::axes_metadata` stores all axes in a std::vector. To represent
+ * which axes each space is 'over' we store those indices. We don't store
+ * the pointers or names for the following reasons:
+ *
+ * * The names of an axis can change after being added. The `nvbench::axes_metadata`
+ * is not aware of the name change, and can't inform this class of it.
+ *
+ * * The `nvbench::axes_metadata` can be deep copied, which would invalidate
+ * any pointers held by this class. By holding onto the index we remove the need
+ * to do any form of fixup on deep copies of `nvbench::axes_metadata`.
+ *
+ *
+ */
+struct iteration_space_base
+{
+  using axes_type = std::vector<std::unique_ptr<nvbench::axis_base>>;
+  using axes_info = std::vector<detail::axis_index>;
+
+  using AdvanceSignature =
+    nvbench::detail::axis_space_iterator::AdvanceSignature;
+  using UpdateSignature = nvbench::detail::axis_space_iterator::UpdateSignature;
+
+  /*!
+   * Construct a new derived iteration_space
+   *
+   * The input_indices and output_indices combine together to allow the iteration space to know
+   * what axi they should query from axes_metadata and where each of those map to in the output
+   * iteration space.
```

Comment:
So `input_indices` determines which elements from the `std::vector<std::unique_ptr<nvbench::axis_base>>` inside `axes_metadata` are referenced by the `iteration_space`? 

What do the `output_indices` reference?

Reply:
> So input_indices determines which elements from the std::vector<std::unique_ptr<nvbench::axis_base>> inside axes_metadata are referenced by the iteration_space?

Correct.

> What do the output_indices reference?

The location in the `detail::state_iterator` vector of results for a given entry in the cross produce of both type and value axis

Reply:
It has been awhile since I wrote this class, and with the removal of adding abitrary zips the `output_indices` might be unneeded if we have the number of type axis. I will see if I can refactor.

Reply:
I have updated the PR and was able to both remove `output_indices` and simplify the interator update logic.